### PR TITLE
Feat/schema diff

### PR DESCRIPTION
**Pull request by heaven00**
Adds tool to query table schema diff using pipeline and table name

- [x] Compare with previous by default
- [x] Add capability to look up schema by version hash when another is provided
- [x] Handle edge case for when there is no schema change
- [x] Manual testing using continue.dev tooling

### CODE REVIEW

File: tests/tools/test_core.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        ]
+    yield [
+        {"id": 1, "name": "Alice", "email": "alice@example.com"},
+        {"id": 2, "name": "Bob", "email": "bob@example.com"},
+    ]
+
+
+def test_get_table_schema_changes_when_schema_has_changed():
+    pipeline_name = "table_schema_change_pipeline"
+    pipeline = dlt.pipeline(pipeline_name, destination="duckdb", dev_mode=True)
```
Comment (zilto) [inline tests/tools/test_core.py]: This will use `duckdb` destination on disk. A trick you can use is

```python
con = duckdb.connect(":memory:")
destination = dlt.destinations.duckdb(con)
pipeline = dlt.pipeline(..., destination=destination)
```

Don't set `dev_mode=True`. It has some quirks that can make debugging harder :/

The rest of the test is great!

File: src/dlt_mcp/_tools/core.py [hunk 2]
```diff
@@ -68,3 +72,56 @@ def get_pipeline_local_state(pipeline_name: str) -> TPipelineState:
...
+
+
+def _get_schemas(another_version_hash, version_hash, dataset) -> pd.DataFrame:
+    if another_version_hash:
+        version_hashes = [version_hash, another_version_hash]
+        # Properly format the list for SQL IN clause
+        quoted_hashes = [f"'{h}'" for h in version_hashes]
+        hashes_str = ",".join(quoted_hashes)
+        return dataset.query(
+            f"select schema from _dlt_version where version_hash in ({hashes_str}) order by inserted_at desc"
```
Comment (zilto) [inline src/dlt_mcp/_tools/core.py]: Should be able to write this as 

```python
(
  dataset.table("_dlt_version")
  .select("schema")
  .where("version", "in", version_hashes)
  .fetchall()
)
``` 

The values should be properly quoted automatically

File: tests/tools/test_core.py (hunk 1 above)
Comment (heaven00) [reply tests/tools/test_core.py]: i did try the ":memory:" but had some issues which I don't recall it will try it out and report back if there are any

File: tests/tools/test_core.py [hunk 3]
```diff
@@ -16,9 +17,14 @@ def user_data(updated_user: bool):
     ]
 
 
-def test_get_table_schema_changes_when_schema_has_changed():
+@pytest.fixture(autouse=True)
+def duckdb_destination():
+    return dlt.destinations.duckdb(duckdb.connect(":memory:"))
```
Comment (zilto) [inline tests/tools/test_core.py]: aah, the `:memory:` solution doesn't work because the tool can't re-attach to the pipeline based on `pipeline_name`. You can revert code to use `dlt.pipeline(..., destination="duckdb")`

Sorry for the inconvenience

File: src/dlt_mcp/_tools/core.py [hunk 4]
```diff
@@ -82,29 +81,30 @@ def get_table_schema_changes(
...
-        return dataset.query(
-            f"select schema from _dlt_version where version_hash in ({hashes_str}) order by inserted_at desc"
-        ).df()
-    return dataset.query(
-        "select schema from _dlt_version order by inserted_at desc limit 2"
-    ).df()
+def _get_schemas(version_hash, another_version_hash, dataset) -> list[str]:
+    table = dataset.table(VERSION_TABLE_NAME).select("schema")
+    version_hashes = [version_hash, another_version_hash]
+    query_with_version_hash = table.where(f"version_hash in {version_hashes}").order_by(
```
Comment (zilto) [inline src/dlt_mcp/_tools/core.py]: we can simplify this code and avoid a potentially costly `order_by` operation by retrieving a single schema at a time

```python
current_schema = pipeline.default_schema
if not other_version_hash:
  other_version_hash = current_schema.previous_hashes[0]

dataset = pipeline.dataset()
query = (
  dataset.table(current_schema.version_table_name)
  .where("version_hash", "=", other_version_hash)
  .select("
)
results = query.fetchall()
other_schema_serialized = results[0]
other_schema = dlt.Schema.from_dict(json.loads(other_schema_serialized))
```

File: src/dlt_mcp/_tools/core.py [hunk 5]
```diff
@@ -68,3 +72,56 @@ def get_pipeline_local_state(pipeline_name: str) -> TPipelineState:
...
+    pipeline_name: str, table_name: str, another_version_hash: Optional[str] = None
+) -> str:
+    """Retrieve the diff between versions of tables compared to it's previous version"""
+    pipeline = dlt.attach(pipeline_name)
+
+    dataset = pipeline.dataset()
+    schemas = _get_schemas(
+        another_version_hash, pipeline.default_schema.version_hash, dataset
+    )
+
```
[the remark this answers is not in the document]
Comment (zilto) [reply src/dlt_mcp/_tools/core.py]: `dlt/common/schema/utils.py` tells me that index 0 is latest -1. Note that it only holds the last 10 hashes, not all of them

```python
def store_prev_hash(
    stored_schema: TStoredSchema, previous_hash: str, max_history_len: int = 10
) -> None:
    # unshift previous hash to previous_hashes and limit array to 10 entries
    previous_hashes = stored_schema["previous_hashes"]
    if previous_hash not in previous_hashes:
        previous_hashes.insert(0, previous_hash)
        if (sur := len(previous_hashes) - max_history_len) > 0:
            del previous_hashes[-sur:]
        # stored_schema["previous_hashes"] = stored_schema["previous_hashes"][:max_history_len]
```

**Review by zilto — COMMENTED** [review]
A few more details about the schema retrieval logic to iron out. I leave a lot of comments to help onboarding you on the code base. Hopefully, it's not too tedious :/ 

> sure I can, but usually I don't test private functions since they are not part of the behavior should I change this assumption overall ? or is it more because you want the diff to be tested out properly?

You're right that private and public interface need different levels of scrutiny. Though, I suggest writing tests for any piece of logic. You never when someone will use the function or modify it. 

In this case, there's a lot happening in `_get_schemas()`. For example, you could check how the function handles a destination with (0, 1, 2, 3) schema versions stored (I saw that you catch `<2` inside the public function). You don't need to add this, but it would benefit from type hints and docstring minimally.
